### PR TITLE
PUB-545 file ingestion

### PR DESCRIPTION
**Pull request by kurt-uk**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/PUB-545

### Change description ###

Overloaded the `uploadPublication` method in the publication controller to take in `MultipartFile` which is then stored in the same blob storage and artefact is created same way assigning the payload url to the artefact payload.

### Headers
3 new headers have been introduced: List type, Court id and Content date

List type is an enum that has been moved from outside of the schema into the headers, this is only needed if the ArtefactType is of LIST or JUDGEMENTS_AND_OUTCOMES. it refers to if the list is a daily list, public list, standard list etc and is an enum.

Court Id has been made a header as it will be needed as flat files are not going to be read to there would be no way of determining what the file was in reference to. This way a file is associated with a court so that it can be pulled back, its a required header for all list types regardless of file or json

Content date is a local date time that the publication refers to, different from the display from/to as these determine when the publication can be shown, this content date refers to what date the publication is referring to so would be used on summary of publications when we say 
```
Bolton Court Daily list 1st January
```
etc

### Validation 
extended the validation service to operate on different Artefact types as these are what determine the different required headers currently

### Exceptions
added `FlatFileIoException` which is thrown if the file could not be converted to input stream, Bad request is returned and asked to check file

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Review by furnivall — APPROVED** [review]
👍

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java [hunk 1]
```diff
@@ -62,6 +59,26 @@ public Artefact createPublication(Artefact artefact, String payload) {
...
+            artefact.getSourceArtefactId(),
+            artefact.getProvenance(),
+            file
+        );
+        artefact.setPayload(blobUrl);
+
+        return artefactRepository.save(artefact);
+    }
+
+    private void checkExisting(Artefact artefact) {
```
Comment (ChrisS1512) [inline src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java:75]: Could we update the method name on this one to highlight it updates the artefact ID, just so it's clear that the method has a side effect of updating the artefact

**Review by ChrisS1512 — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java [hunk 2]
```diff
@@ -93,6 +130,76 @@ public PublicationController(PublicationService publicationService) {
...
     }
 
+    /**
+     * This endpoint takes in the Artefact, which is split over headers and also the payload flat file.
+     *
+     * @param provenance       Name of the source system.
+     * @param sourceArtefactId Unique ID of what publication is called by source system.
+     * @param type             List / Outcome / Judgement / Status Updates.
+     * @param sensitivity      Level of sensitivity.
+     * @param language         Language of publication.
```
[the remark this answers is not in the document]
Comment (kurt-uk) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java:140]: 👍🏼

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java (hunk 1 above)
Comment (kurt-uk) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java]: 👍🏼

**Review by furnivall — DISMISSED** [review]
👍

**Review by ChrisS1512 — DISMISSED** [review]

**Review by furnivall — APPROVED** [review]
👍

**Review by ChrisS1512 — APPROVED** [review]

**Review by ChrisS1512 — DISMISSED** [review]

**Review by alberto-hortiguela — DISMISSED** [review]

**Review by ChrisS1512 — APPROVED** [review]

**Review by furnivall — APPROVED** [review]
👍

**Review by ChrisS1512 — APPROVED** [review]

**Pull request merged**